File: reference/aeonscript/encoder.py
"""High-level AeonScript encoder.

Pipeline (top-to-bottom):
    payload (bytes)
      → L5 semantic tag prepended
      → L4 Reed-Solomon error correction applied
      → L1 DNA encoding (with constraint-aware base permutation)
      → split into oligos of configured length

Each oligo is a self-contained, synthesizable DNA string.
"""

from __future__ import annotations

import hashlib
import os
from pathlib import Path
from typing import Iterable

from .error_correction import encode_with_ecc
from .physical import bytes_to_dna
from .scrambler import scramble
from .semantic import make_tag

DEFAULT_OLIGO_LENGTH = 200  # bases; SPEC §2.4 recommendation
DEFAULT_RS_PARITY = 32  # SPEC §3.1
DEFAULT_RS_BLOCK = 223


_FORBIDDEN_TAG_CHARS = (";", "=", "|")


def _validate_tag_value(name: str, value: str) -> None:
    """Raise ValueError if `value` contains characters reserved by the L5 wire format."""
    bad = [c for c in _FORBIDDEN_TAG_CHARS if c in value]
    if bad:
        raise ValueError(
            f"L5 tag value for {name} contains reserved characters {bad!r}: "
            f"{value!r}. Per SPEC-v0.1 §4.5, the characters ';', '=', '|' "
            f"are not permitted in tag values (escape sequences reserved for v0.2)."
        )


def _blake3_or_sha256(data: bytes) -> str:
    """Return a hash of the data. Uses BLAKE3 if available, else SHA-256."""
    try:
        import blake3  # type: ignore

        return f"blake3:{blake3.blake3(data).hexdigest()}"
    except ImportError:
        return f"sha256:{hashlib.sha256(data).hexdigest()}"


def encode_bytes(
    data: bytes,
    *,
    mime_type: str = "application/octet-stream",
    block_id: str | None = None,
    oligo_length: int = DEFAULT_OLIGO_LENGTH,
    rs_parity: int = DEFAULT_RS_PARITY,
    rs_block_size: int = DEFAULT_RS_BLOCK,
    extra_tag_fields: dict[str, str] | None = None,
) -> list[str]:
    """Encode arbitrary bytes into a list of DNA oligos.

    Returns
    -------
    list[str]
        Each element is a DNA sequence (A/C/G/T) of approximately
        `oligo_length` bases, ready for synthesis.
    """
    if block_id is None:
        block_id = _blake3_or_sha256(data)[:24]  # short prefix

    # ---- L5: prepend semantic tag --------------------------------------
    # The L5 wire format uses ';' as the field separator and '=' as the
    # key/value separator. In v0.1 these characters are forbidden in any
    # value (escape sequences are reserved for v0.2). We reject loudly
    # rather than silently truncate — losing the 'charset=utf-8' part of a
    # MIME type would be a silent data corruption.
    _validate_tag_value("mime_type", mime_type)
    if extra_tag_fields:
        for k, v in extra_tag_fields.items():
            _validate_tag_value(f"extra_tag_fields[{k!r}]", v)

    tag = make_tag(
        mime_type=mime_type,
        payload_len=len(data),
        block_id=block_id,
        CHECKSUM=_blake3_or_sha256(data),
        **(extra_tag_fields or {}),
    )
    tagged = tag.encode("ascii") + data

    # ---- L4: Reed-Solomon error correction -----------------------------
    rs_encoded = encode_with_ecc(tagged, nsym=rs_parity, block_size=rs_block_size)

    # ---- Scrambler: uniformize byte distribution ----------------------
    rs_encoded = scramble(rs_encoded)

    # ---- L1: DNA encoding with constraint enforcement -----------------
    # We chunk the RS-encoded bytes so each resulting oligo fits oligo_length
    # bases (each byte = 4 bases, plus 4-base permutation prefix per oligo).
    bytes_per_oligo = (oligo_length - 4) // 4
    if bytes_per_oligo < 16:
        raise ValueError(f"oligo_length={oligo_length} too small")

    oligos = []
    for i in range(0, len(rs_encoded), bytes_per_oligo):
        chunk = rs_encoded[i : i + bytes_per_oligo]
        oligos.append(bytes_to_dna(chunk))

    return oligos


def encode_file(
    path: str | os.PathLike,
    *,
    mime_type: str | None = None,
    **kwargs,
) -> list[str]:
    """Encode a file from disk into DNA oligos."""
    p = Path(path)
    if not p.exists():
        raise FileNotFoundError(p)

    if mime_type is None:
        # Naive MIME guess from extension
        ext_map = {
            ".txt": "text/plain",
            ".md": "text/markdown",
            ".json": "application/json",
            ".jpg": "image/jpeg",
            ".jpeg": "image/jpeg",
            ".png": "image/png",
            ".pdf": "application/pdf",
            ".bin": "application/octet-stream",
        }
        mime_type = ext_map.get(p.suffix.lower(), "application/octet-stream")

    data = p.read_bytes()
    block_id = kwargs.pop("block_id", p.name)
    return encode_bytes(
        data,
        mime_type=mime_type,
        block_id=block_id,
        **kwargs,
    )

File: reference-rs/src/encoder.rs
//! High-level AeonScript encoder.
//!
//! Pipeline (top-to-bottom):
//!
//! 1. L5 — prepend the semantic tag.
//! 2. L4 — apply Reed-Solomon error correction (skeleton in v0.1, passes through).
//! 3. Scrambler — xorshift32 PRBS to balance GC content.
//! 4. L1 — encode bytes as DNA with constraint-aware base permutations.
//! 5. Chunk into oligos of the configured target length.

use crate::error_correction::{encode_with_ecc, EccError, DEFAULT_BLOCK, DEFAULT_NSYM};
use crate::physical::{bytes_to_dna, PhysicalError};
use crate::scrambler::scramble;
use crate::semantic::{make_tag, validate_tag_value, TagError};

/// Default target oligo length in bases, per SPEC §2.4.
pub const DEFAULT_OLIGO_LENGTH: usize = 200;

/// Errors produced by the encoder.
#[derive(Debug, thiserror::Error)]
#[allow(missing_docs)]
pub enum EncodeError {
    #[error("L1 encoding failed: {0}")]
    Physical(#[from] PhysicalError),
    #[error("L4 RS encoding failed: {0}")]
    Ecc(#[from] EccError),
    #[error("L5 tag error: {0}")]
    Tag(#[from] TagError),
    #[error("Oligo length too small (minimum is 64 bases, got {0})")]
    OligoTooSmall(usize),
}

/// Encode arbitrary bytes into a list of DNA oligos.
///
/// Each returned [`String`] is a synthesizable DNA sequence (only `A`,
/// `C`, `G`, `T`), of approximately `DEFAULT_OLIGO_LENGTH` bases.
pub fn encode_bytes(
    data: &[u8],
    mime_type: &str,
    block_id: &str,
) -> Result<Vec<String>, EncodeError> {
    encode_bytes_full(data, mime_type, block_id, DEFAULT_OLIGO_LENGTH)
}

/// Encode bytes with an explicit oligo target length.
pub fn encode_bytes_full(
    data: &[u8],
    mime_type: &str,
    block_id: &str,
    oligo_length: usize,
) -> Result<Vec<String>, EncodeError> {
    if oligo_length < 64 {
        return Err(EncodeError::OligoTooSmall(oligo_length));
    }

    // L5: validate then build the semantic tag
    validate_tag_value("mime_type", mime_type)?;
    validate_tag_value("block_id", block_id)?;
    let tag = make_tag(mime_type, data.len(), block_id);

    let mut tagged = tag.into_bytes();
    tagged.extend_from_slice(data);

    // L4: Reed-Solomon RS(255, 223) over GF(256)
    let rs_encoded = encode_with_ecc(&tagged, DEFAULT_NSYM, DEFAULT_BLOCK)?;

    // Scrambler: uniformise byte distribution
    let stream = scramble(&rs_encoded);

    // L1: chunk into oligos
    let bytes_per_oligo = (oligo_length - 4) / 4;
    let mut oligos = Vec::new();
    for chunk in stream.chunks(bytes_per_oligo) {
        oligos.push(bytes_to_dna(chunk)?);
    }
    Ok(oligos)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::decoder::decode_oligos;

    #[test]
    fn round_trip_short_text() {
        let data = b"AeonScript round-trip test.";
        let oligos = encode_bytes(data, "text/plain", "t1").unwrap();
        let decoded = decode_oligos(&oligos).unwrap();
        assert_eq!(decoded.as_slice(), data);
    }

    #[test]
    fn rejects_mime_with_semicolon() {
        let result = encode_bytes(b"x", "text/plain; charset=utf-8", "t1");
        assert!(matches!(result, Err(EncodeError::Tag(TagError::ReservedCharacters { .. }))));
    }
}
